plasmapy/tests/helpers/tests/test_expected.py

```diff
@@ -0,0 +1,178 @@
+
+import collections
+import pytest
+
+from plasmapy.tests.helpers.expected import (
+    _is_exception,
+    _is_warning,
+    _is_warning_and_value,
+    ExpectedTestOutcome,
+)
+from plasmapy.utils.exceptions import PlasmaPyWarning
+
+expected_exception = KeyError
+expected_warning = UserWarning
+expected_value = 42
+
+Case = collections.namedtuple("Case", ["argument", "attribute", "correct_outcome"])
+
+expected_value_and_warning = (expected_value, expected_warning)
+expected_warning_and_value = (expected_warning, expected_value)
+
+expected_outcome_test_cases = [
+    Case(expected_exception, "expected_exception", expected_exception),
+    Case(expected_exception, "expected_outcome", expected_exception),
+    Case(expected_exception, "expecting_an_exception", True),
+    Case(expected_exception, "expecting_a_warning", False),
+    Case(expected_exception, "expecting_a_value", False),
+    Case(expected_warning, "expected_warning", expected_warning),
+    Case(expected_warning, "expected_outcome", expected_warning),
+    Case(expected_warning, "expecting_an_exception", False),
+    Case(expected_warning, "expecting_a_warning", True),
+    Case(expected_warning, "expecting_a_value", False),
+    Case(expected_value, "expected_value", expected_value),
+    Case(expected_value, "expected_outcome", expected_value),
+    Case(expected_value, "expecting_an_exception", False),
+    Case(expected_value, "expecting_a_warning", False),
+    Case(expected_value, "expecting_a_value", True),
+    Case(expected_value_and_warning, "expected_warning", expected_warning),
+    Case(expected_value_and_warning, "expected_value", expected_value),
+    Case(expected_value_and_warning, "expecting_an_exception", False),
+    Case(expected_value_and_warning, "expecting_a_warning", True),
+    Case(expected_value_and_warning, "expecting_a_value", True),
+    Case(expected_value_and_warning, "expected_outcome", expected_warning_and_value),
+    Case(expected_warning_and_value, "expected_warning", expected_warning),
+    Case(expected_warning_and_value, "expected_value", expected_value),
+    Case(expected_warning_and_value, "expecting_an_exception", False),
+    Case(expected_warning_and_value, "expecting_a_warning", True),
+    Case(expected_warning_and_value, "expecting_a_value", True),
+    Case(expected_warning_and_value, "expected_outcome", expected_warning_and_value),
+]
+
+
+@pytest.mark.parametrize("case", expected_outcome_test_cases)
+def test_expected_test_outcome_attributes(case: Case):
+    """Test that `ExpectedTestOutcome` attributes each return the expected values."""
+    expected_outcome = ExpectedTestOutcome(case.argument)
+    result = expected_outcome.__getattribute__(case.attribute)
+    if result is not case.correct_outcome and result != case.correct_outcome:
+        pytest.fail(
+            f"ExpectedTestOutcome({case.argument}) results in {repr(result)} "
+            f"but should result in {case.correct_outcome}."
+        )
```

Comment:
I keep running into these blocks. I honestly don't think they're helpful.

Consider this basic test:


```python
import pytest

@pytest.mark.parametrize("answer", [42, 44])
def test_answer_to_everything(answer):
    assert 41 < answer < 43, "Answer to everything out of bounds (cf. Douglas, 1979)"
```

When I run this, I get 

```
🕙[ 11:09:22 ] ❯ pytest test_pytest.py                                                                                                                                  [1]
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.9.2, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
Matplotlib: 3.4.0
Freetype: 2.10.4
rootdir: /home/dominik
plugins: datadir-1.3.1, profiling-1.7.0, mpl-0.12, ordering-0.6, regressions-2.2.0, pudb-0.7.0, testmon-1.0.3, hypothesis-6.3.4, cov-2.11.1, anyio-2.2.0, Faker-6.6.3, mock-3.1.1, xdist-2.2.1, forked-1.3.0, black-0.3.12, xonsh-0.9.24
collected 2 items

test_pytest.py .F                                                                                                                                                    [100%]

================================================================================= FAILURES =================================================================================
______________________________________________________________________ test_answer_to_everything[44] _______________________________________________________________________

answer = 44

    @pytest.mark.parametrize("answer", [42, 44])
    def test_answer_to_everything(answer):
>       assert 41 < answer < 43, "Answer to everything out of bounds (cf. Douglas, 1979)"
E       AssertionError: Answer to everything out of bounds (cf. Douglas, 1979)
E       assert 44 < 43

test_pytest.py:5: AssertionError
========================================================================= short test summary info ==========================================================================
FAILED test_pytest.py::test_answer_to_everything[44] - AssertionError: Answer to everything out of bounds (cf. Douglas, 1979)
======================================================================= 1 failed, 1 passed in 0.61s ========================================================================
```

The values are shown and I didn't need to hand write any f-strings.

So I would prefer to change those (and stick to this style in our tests, and ensure the same is stated in the documentation guide) to *(trying to invert your condition...)*... and the reasoning is obvious enough from the simple test, so the message is unnecessary...

```suggestion
    assert result is case.correct_outcome
    assert result == case.correct_outcome
```

And now I honestly wonder what the point is of having both `is` and `==` checks. That sounds off. I haven't looked into the class itself yet, just trying to familiarize myself with how it works on the surface level; but if the correct outcome of a test were to be a string...

```python
[ins] In [3]: A = "Engage."

[ins] In [4]: B = "Engage!"

[ins] In [5]: A[:-1] == B[:-1]
Out[5]: True

[ins] In [7]: A[:-1] is B[:-1]
Out[7]: False
```

That latter condition does not sound all that helpful. Are you 100% sure we need it?